### ISSUE TITLE
ZonedDateTime serialization error (Version: 2.4.1)

### ISSUE THREAD

**Issue by ClaudiuGuja** [kobylynskyi/graphql-java-codegen#260]
Issue: ZonedDateTime string not escaped.

When sending a variable through graphql of type ZonedDateTime, in `GraphQLRequestSerializer` method `getEntry`, the ZonedDateTime goes through the last else branch, and toString() is called on it. This leads to a request that does not fit the JSON format and a BadRequest exception is thrown.

Instead, when sending a string that contains the result of toString() called on the ZonedDateTime, the string is escaped at line 100 of GraphQLRequestSerializer, and the call is made correctly.

Possible solutions: 
- add another branch to the getEntry method
- ability to inject a custom ObjectMapper into GraphQLRequestSerializer. Sadly this means no static methods and brings with it another host of problems

**Comment by ClaudiuGuja** [comment]
I have also tried to use the custom mapping options
```
customTypesMapping = [
            ZonedDateTime: "java.time.ZonedDateTime"
    ]
    customAnnotationsMapping = [
            ZonedDateTime: "@com.fasterxml.jackson.databind.annotation.JsonSerialize(using = smf.stui.graphqlclient.config.CustomSerializer.class)"
    ]
```
the CustomSerializer having the method
```
 @Override
    public void serialize(ZonedDateTime value, JsonGenerator gen, SerializerProvider provider) throws IOException {
        gen.writeString("anything");
    }
```
but still returning the same error, as if the serialization is not applied

**Comment by kobylynskyi** [comment]
@ClaudiuGuja as of today it is not supported. Can you please try adding a custom type mapping for ZonedDateTime?
`ZonedDateTime: String`

**Comment by ClaudiuGuja** [comment]
Solution I went with:
- Make a JsonSerializationService that can be injected with an ObjectMapper
- Have the method of `toHttpJsonBody` with it's current functionality, and in `getEntry` instead of
```
else {
            return input.toString();
        }
```
I've added 
```
{
            try {
                return objectMapper.writeValueAsString(input);
            } catch (JsonProcessingException e) {
                throw new UnableToBuildJsonQueryException(e);
            }
        }
```

I would like to ask though if the customAnnotationMappings are only for the results, and not for the requests, or if there are plans in the future to add the posibility of formatting on the query requests as well

**Comment by kobylynskyi** [comment]
@ClaudiuGuja annotations are being added to all types, so `ObjectMapper` inside `GraphQLRequestSerializer` should pick it up.
Can you please show your schema excerpt which you have problems with?

**Comment by ClaudiuGuja** [comment]
@kobylynskyi 

build.gradle config:
```
customTypesMapping = [
            ZonedDateTime: "java.time.ZonedDateTime"
    ]
    customAnnotationsMapping = [
            ZonedDateTime: "@SomeAnnotation"
    ]
```

schema.graphqls
```
type Query {
    queryINeed(
        input: ZonedDateTime!
    ): [ResponseContainingDate]
    otherQueryUsedAsExamle(
        input: InputContainingDate!
    ): String
}

type ResponseContainingDate{
    a: ZonedDateTime
}

input InputContainingDate{
    a: ZonedDateTime
}

# ZonedDateTime Scalar
scalar ZonedDateTime
```

Conclusions from the example above:
indeed the annotation is added to the generated classes `InputContainingDate`, and `ResponseContainingDate`, but it's not added to `QueryINeedRequest`. 

But even so, both `OtherQueryUsedAsExampleRequest` and `QueryINeedRequest` are have `toHttpJsonBody` called on them. When I tried following the serialization, I saw that `getEntry(inputEntry.getValue, jsonQuery)` is called of them, and inside, `toString` is called on both. so that they are both serialized wrongly (not in JSON format), not through the ObjectMapper.

**Comment by kobylynskyi** [comment]
The thing is that if we use `OBJECT_MAPPER.writeValueAsString(input);` then the query is being serialized not in GraphQL-style:
<img width="949" alt="image" src="https://user-images.githubusercontent.com/2664706/88683871-e653cc00-d0b9-11ea-9afe-88b1ca4943d1.png">

So we need to find a way how to serialize the whole GraphQLRequest properly.

**Comment by kobylynskyi** [comment]
As of today, I would suggest you to add a custom mapping from `ZonedDateTime` to `String`:
`customTypesMapping = [ "ZonedDateTime": "String" ]`
But this will impact all other places where you might want to keep as `java.time.ZonedDateTime`.

As another solution, I can add a possibility to specify custom type mapping for operation input:
`customTypesMapping = [ "queryINeed.input": "String" ]`
If you are ok with this, I will release it today/tomorrow as part of v2.4.2.

In both scenarios you can pass a serialized value of ZonedDateTime in the format that you wish (while you are constructing a GraphQLRequest object):
```java
QueryINeedQueryRequest request = new QueryINeedQueryRequest.Builder()
        .setInput(ZonedDateTime.now().format(DateTimeFormatter.ISO_DATE_TIME))
        .build();
```

**Comment by ClaudiuGuja** [comment]
Indeed, just using the ObjectMapper there doesn't solve the whole problem. I'd like to think a bit more on how this might be solved before applying a fix. At the moment I can do by with just using the ObjectMapper, but if the input is a more complex object, then the JSON format is not good enough anymore

**Comment by ClaudiuGuja** [comment]
@kobylynskyi , how about the following solution: 
- add a marker interface to the input types
- have getEntry() check if the input is of the marker interface type. If so, then use toString() on it. if not, then ObjectMapper.

This way, the input field names can still be formatted to fit with graphql, and the values should be ok if they pass through the objectMapper (tested for example on Int, to make sure that it doesn't add the quotation marks to it)

**Comment by ClaudiuGuja** [comment]
That would still leave the issue of ZonedDateTimes that are directly in the request, since the request keeps it's parameters in a Map, and no annotations can be added there. For that, perhaps a getter method for the ObjectMapper from GraphQLRequestSerializer, so that people can customize it at the start of their application running.

Thoughts?

**Comment by kobylynskyi** [comment]
@ClaudiuGuja yes, with this approach we will still have the issue with custom types serialization.
Maybe we need to introduce another config taking a list of types for which we delegate the serialization to ObjectMapper instead of a toString(). But then we would need to somehow make ObjectMapper injectable (or at least configurable)

**Comment by jxnu-liguobin** [comment]
Hi,are you talking about escaping parameters, like this id parameter
```
{
    "query": "query { human(id: \"1001\"){ name id appearsIn friends { name id appearsIn friends { name id appearsIn secretBackstory __typename } secretBackstory __typename } homePlanet secretBackstory email __typename } }"
}
```
Can we support the following format? 
```
{
    "operationName": "createUserVariable",
    "variables": {
        "userVariable": {
            "name": "testgraphql",
            "key": "test_132",
            "valueType": "int",
            "description": "132"
        }
    },
    "query": "mutation createUserVariable($userVariable: VariableInput!) {\n  createUserVariable(userVariable: $userVariable) {\n    name\n    __typename\n  }\n}\n"
}
```

**Comment by kobylynskyi** [comment]
@jxnu-liguobin, there are no issues with escaping parameters.
The issue is that `GraphQLRequestSerializer` serializes everything using `toString()` (which is fine in most cases because it is overridden by all classes generated by this plugin). But if custom non-generated class is supplied (like `ZonedDateTime`), then `ZonedDateTime.toString()` returns it's own default format (e.g.: `2020-07-30T22:17:17.884-05:00[America/Chicago]`) which is not what we want.

**Comment by jxnu-liguobin** [comment]
Oh, it does seem like a common scene.

**Comment by ClaudiuGuja** [comment]
@kobylynskyi I don't think it's necessarily a good practice, but by having a getter for the static ObjectMapper inside GraphQLRequestSerializer, we could configure it. (for example through a piece of code that runs at startup)

**Comment by kobylynskyi** [comment]
@ClaudiuGuja yes, we could do that. But still, there should be a possibility to specify which fields to serialize using `toString()` or `ObjectMapper`.

**Comment by ClaudiuGuja** [comment]
@kobylynskyi , can you please tell me how you made the projection classes extend GraphQLResponseProjection? we're still looking for a nested inputs solution, and we were wondering how to add a marker interface to the inputs, so we can diferentiate the generated ones from the Java native objects, such as ZonedDateTime.

Thank you

**Comment by circap-ppb** [comment]
Hi @kobylynskyi, I have a quick fix for this issue, but I'm not able to push my changes into github.

The solution is basically, change this method on the **GraphQLRequestSerializer.java**:
`    public static String getEntry(Object input) {
        if (input == null) {
            return null;
        }
        if (input instanceof Collection<?>) {
            Collection<?> inputCollection = (Collection<?>) input;
            return inputCollection.stream()
                    .map(GraphQLRequestSerializer::getEntry)
                    .collect(Collectors.joining(", ", "[ ", " ]"));
        }
        if (input instanceof Enum<?>) {
            return input.toString();
        } else if (input instanceof String) {
            return escapeJsonString(input.toString());
        } **else if (input instanceof ZonedDateTime) {
            return "\"" + input.toString() + "\"";
        }** else {
            return input.toString();
        }
    }`

**Comment by kobylynskyi** [comment]
@ClaudiuGuja all classes generated by graphql-java-codegen are annotated with `@Generated`.

@circap-ppb thanks for your input. But the thing is we can't specify all possible types that should be serialized in such a way (using `toString()`). There could be an unlimited amount of such classes, so we need to find a generic way to do so.

**Comment by kobylynskyi** [comment]
@ClaudiuGuja how about the following solution:
Fields that should be serialized using `ObjectMapper` will be defined as following:
```groovy
customTypesMapping = [
    ZonedDateTime: "java.time.ZonedDateTime"
]
useObjectMapperForRequestSerialization = [
    "ZonedDateTime"
]
```
So that generated class that has specified graphql type will have slightly different `toString()` method:
```java
joiner.add("yourOldFashionedField: " + GraphQLRequestSerializer.getEntry(yourOldFashionedField));
joiner.add("yourZoneDateTimeField: " + GraphQLRequestSerializer.getEntry(yourZoneDateTimeField, true));
```
And `GraphQLRequestSerializer` will have the overloaded `getEntry` method:
```java
public static String getEntry(Object input, boolean useObjectMapper) {
    ...
    // all code remains as-is except for a new condition:
    if (useObjectMapper) {
        try {
            return OBJECT_MAPPER.writeValueAsString(input);
        } catch (JsonProcessingException e) {
            throw new UnableToBuildJsonQueryException(e);
        }
    }
    // ...
}
```

**Comment by ClaudiuGuja** [comment]
It does seem like it would solve our problem

**Comment by kobylynskyi** [comment]
The change described in [my previous comment](https://github.com/kobylynskyi/graphql-java-codegen/issues/260#issuecomment-725199906) will be released today as part of graphql-java-codegen 4.0.0
